src/core/kernel/kernel.rs

```diff
@@ -0,0 +1,239 @@
+use std::fmt;
+
+use anyhow::{bail, Result};
+
+use super::{inspect, symbols};
+
+/// Kernel symbol representation. Only supports traceable symbols: events and
+/// functions.
+pub(crate) struct Symbol {
+    r#type: SymbolType,
+    name: String,
+}
```

Comment:
Looking at how it's used and what I'm working on for usdt/uprobe, what do you think about this:

```rust
struct TracePointSymbol {
   provider: String,
   name: String,
}
enum Symbol {
    Kprobe(String),
    Tracepoint(TracePointSymbol)
    // Later
    USDT(USDTSymbol)
} 
// Later
struct USDTSymbol {
    pid: pid_t,
    path: PathBuf,
    provider: String
    name: String
}
```
Having userspace probes share the Symbol representation will make the APIs much more unified.

Reply:
I think my problem is that `Symbol` and `Target` are, as you argue, two conceptually different things but they seem interleaved.

collectors can attach to `Symbol`s or `Target`s
```rust
impl Kernel {
[...]
pub(crate) fn register_hook_to_symbol(
        &mut self,
        hook: Hook,
        r#type: ProbeType,
        symbol: Symbol,
    ) -> Result<()>
    
  pub(crate) fn register_hook_to(&mut self,
        hook: Hook,
        r#type: ProbeType,
        target: &str,
    ) -> Result<()> {
        let symbol = Symbol::from_name(target)?;
        self.register_hook_to_symbol(hook, r#type, symbol)
    } 
}
```

When `Kernel` sends the request down to probes they receive a `Symbol`.
```rust
impl ProbeBuilder for KprobeBuilder {
fn attach(&mut self, symbol: &Symbol, _: &TargetDesc) -> Result<()>
}
impl ProbeBuilder for RawTracepointBuilde {
fn attach(&mut self, symbol: &Symbol, _: &TargetDesc) -> Result<()>
}
```

But it's possible that a probe type cannot be added to a symbol, so `Symbol` has a `r#type` (which for me is a strong indicator that can be made into a rust variant) and a `can_probe` that yields a 1:1 relationship between `Symbol.type` and `TargetType`. Also, the `Symbol.attach_name()` depends on the `Symbol.type`, ultimately on the `TargetType`.

So for me a more accurate data model would be:
```rust
enum Target {
    Kprobe(Kprobe)
    Tracepoint(Tracepoint)
    USDT(USDT)
    Uprobe(Uprobe)
}

impl Target {
   /// From strings such as: "kprobe::func_name" (or just "func_name"), "tp::name", "usdt:provider:name", "uprobe::func_name".
   fn from_string(str) -> Result<Self> 
   /// We may want to get the symbol from the target. Some targets (e.g: USDT) might not have symbols associated!
   fn get_symbol(&self) -> Result<Option<Symbol>>
}

enum Symbol {
   Kernel(KernelSymbol)
   User(UserSymbol)
}

```

Collectors can still use `Symbol` to inspect features on kernel or userspace programs. We can even build _concrete_ `Target`s from `Symbol`s if we want (Kprobe.from_symbol(KernelSymbol)). However, the probe attachment API (a.k.a `ProbeBuilder` would just receive a `Target` which has all the information we need to attach to it.

I'm not saying this should be done in this PR but I may need to do this in order to add USDT support. WDYT?

Reply:
We can make `Target` a trait instead but ultimately `ProbeBuilder`s will have to `downcast` them (or `match` them)

Reply:
I think we're getting somewhere :)

> What would be the advantage of this? I believe giving a (ProbeType, Symbol) tuple or a Target is exactly the same, as a Target is a Symbol associated with a ProbeType. Is there a specific blocking point for USDT support here? Is it we can attach per-target data using this approach?

I guess my initial inclination towards Symbol -> Target rather than the opposite is that I didn't see USDT as a symbol, i.e: it's not a function. But if I mentally relax the meaning of Symbol then I think it's OK.


> 
> enum Symbol {
>     Kernel(KernelSymbol),
>     Userspace(UserspaceSymbol),
> }

Is `KernelSymbol` itself an enum with `Function` and `Tracepoint` variants?

> trait SymbolTrait {
>     fn name(&self) -> &str;
>     fn attach_name(&self) -> &str;
>     fn addr_name(&self) -> &str;
>     fn typedef_name(&self) -> &str;
>     fn addr(&self) -> Result<u64>;
>     fn nargs(&self) -> Result<u32>;
>     fn parameter_offset(&self, parameter_type: &str) -> Result<Option<u32>>;
>     [...]
>     // With the following we can remove `can_probe()`
>     fn to_probe(&self, r#type: ProbeType) -> Result<Probe>;
> }

If all variants need to implement all of these we might as well use a `dyn Symbol` here but I think User and Kernel symbols might be different enough to fit into a clean trait. E.g: `typedef_name` does not make sense for USDT, all userspace symbols need `pid` and `path` which doesn't make sense for kernel, etc.

Alternatively, if things like `name`, `attach_name`, `addr_name`, etc are going to be used inside each `ProbeBuilder` they could be implemented only if needed on each variant since each `ProbeBuilder` would have to `match` the concrete `Probe` type variant type anyhow.

Finally, are you planning to do this in this PR or should I do this rework as part of the USDT work?

Reply:
> I feel there's something not quite right about this. It looks to me the `Symbol` now would have two variant not sharing much, and not used in the same code paths. To split shared helpers from specific ones we can have a trait + variant specific implementations, but that comes with a cost when using those.
> 
> Or I might be a bit confused with all the changes :) I have a few questions:
> 
>     * Where do you see the `Symbol` obj being used, where both variants makes sense?
> 

In all the `ProbeSet` / `ProbeBuilder` APIs. I'm trying to avoid having an almost identical copy of `core/probe/kernel/kernel.rs`.

>     * What code could be reused between the two `Symbol` variants?
> 

The way I see it, `enums` are not about sharing code but about sharing APIs. Conceptually, user probes also have `Target`s or `Probe`s, which also have `Symbols`, and also need `ProbeBuilder`s to attach `Hooks` to them. The alternative to create uprobe/usdt support is either duplicating those structs, or duplicating the functions on those structs.

Reply:
Maybe I should just start by duplicating everything and we can later see if we can refactor something in some common part.